### ISSUE TITLE
Running ‘prism proxy’ with –errors flag doesn’t enforce required header

### ISSUE THREAD

**Issue by TylerNielsen** [stoplightio/prism#1091]
**Describe the bug**
I’m running prism as a proxy for another service, and the OpenAPI spec for the service specifies that an x-authorization header is required.

When I submit the request to the proxy, I see that prism is showing a validation violation due to missing header. However, I’m getting the response from the “live” service itself, rather than the error about the missing header from prism proxy. I’m using the ‘–errors’ flag. I’ve also confirmed that other sorts of errors ARE reported.

**To Reproduce**
Requires Docker
1. Clone the following repository: [TylerNielsen/prism-repro-issue](https://github.com/TylerNielsen/prism-repro-issue). I don't have a running service that I can share to reproduce this issue, so this repro will allow a docker container that can be spun up as a target. Note that this repo also contains the example OpenApi.json file.
2. Build the docker image from the cloned repo: `docker build -t prism-repro-1212 -f Dockerfile .`
3. Run the mock service in a container: `docker run -it --rm -p 8080:8080 prism-repro-1212`
4. Start the prism proxy by running this command: `prism proxy openapi.json http://127.0.0.1:8080 --errors`
5. Run the following command: `curl -X POST -H "Content-Type: application/json" http://0.0.0.0:4010/v1/presigned-url -d "{\"test\": \"this is a test\"}"`
6. See that the response is returned from the "live" service, but prism is throwing an error about the missing "x-company-auth" header.

**Expected behavior**
Since the `--errors` flag is included, the prism proxy service should reject the incoming request and return an error like it does for missing required properties in the body.

**Environment (remove any that are not applicable):**
 - Library version: 3.3.3
 - OS: Windows 10
 - Browser: N/A

**Additional context**
If you run prism as a mock server, rather than as a proxy, I see that the request without the x-company-auth header is rejected.

**Comment by XVincentX** [comment]
Thanks for the great repro steps @TylerNielsen , will review!

**Comment by XVincentX** [comment]
@TylerNielsen I've dug a little bit into this and the behaviour to me appears to be correct.

Fundamentally Prism has been able to recognise that you're missing a required request header, as you can see from the log statements.

What happens now is that when mocking (not your case) — the negotiator, based on the failed request validation, will look for an error response or generate one if no-one is specified in the document.

On the other hand, when acting like a proxy (which is your case), Prism will not take any decision based on the input and will forward the request to the upstream server no matter what.

The application contained in the repository you've provided is not — and thus Prism is just returning such response to your client. In a certain sense, we can say that Prism has revealed a discrepancy between the contract file and the implementation :)

Keep in mind that the `--errors` flag is only acting on the response (body and headers) where Prism has enough information to know whether the returned content is legit or not. 

It won't act on the request, although it will still tell you that what's going to the upstream server is wrong.

TL; DR

1. `--errors` acts only on response violations, not on request validations
2. Prism will always forward to the upstream server when acting in Proxy mode, no matter the request validation errors.

Does that clarify?


//cc @philsturgeon Please check this out — it's good to have a confirmation (from a product perspective) here.

**Comment by TylerNielsen** [comment]
@XVincentX - I appreciate you looking into it so quickly. If that's the way it's coded, that's the way it's coded. However I would say as a consumer, that seems to contradict or at least be missing from the [documentation](https://stoplight.io/p/docs/gh/stoplightio/prism/docs/guides/03-validation-proxy.md?srn=gh/stoplightio/prism/docs/guides/03-validation-proxy.md).

From the documentation:

> As of Prism v3.2, API consumers can funnel their development traffic through Prism running as proxy, and then relay that traffic to the API-in-progress. It will report any mistakes it notices along the way, **either with the requests you're sending or the responses coming back from the server.**
> In this use case, the OpenAPI documents would be provided by the API team, distributed to the API consumers, and the proxy server would be pointing to the dev environment. We'd also want to enable --errors so Prism **will blow up your request if a mismatch is noticed.**

The above doesn't seem to imply that --errors would only throw on response. It seems to specifically call out it works on both.  

Personally, I'd definitely want to know whether either the request or response is incorrect. If the API is in progress, there otherwise is a case that the actual implementation isn't correctly enforcing the contract, and neither consumer or implementation may know. We could look at prism output, but I want something even more aggressive and hard to miss - I want the request to fail. Part of the power of prism could be ensuring that the actual implementation is actually respecting the rules it says it will from the request side of things. 

Question #1  is: are you sure that --errors is only  intended to act on response given the pasted documentation?

Question #2: If that is the intended behavior, can the documentation be updated to call that out? 

Question #3: Thoughts on allowing this to be configurable so prism can throw errors on either request, response, or both?

Question #4: If #3 isn't possible, is there a way for prism to output it's validation errors in a consumable way so we can wrap our own way of checking for these violations automatically?

**Comment by philsturgeon** [comment]
@TylerNielsen sorry for the confusion here, there has certainly been some communication. I wrote the gitub issues to plan the work and wrote the documentation, so it's either way its my fault. Let me chat with the team to see what the best next step is!

**Comment by XVincentX** [comment]
@TylerNielsen Sorry for the confusion :)

We had an internal discussion and there's been a misunderstanding in the way the code should behave so yes, you're definitely right. We'll put this on top of our things to do and get it done as soon as possible.

Thanks for the patience!

**Comment by TylerNielsen** [comment]
@XVincentX and @philsturgeon - Appreciate the quick follow-ups, I don't feel like I've had to be patient : ) 

I just started playing around prism, and am excited about the potential - thanks!

**Comment by XVincentX** [comment]
@TylerNielsen Can you try out fix/request-check-in-proxy (#1099) and let me know if that solves your use case? It works with the demo, but I was wondering whether you don't have a real use case you can try that out with.

**Comment by TylerNielsen** [comment]
@XVincentX - Was there a comment that got deleted? I got an email about you asking for me to checkout PR #1099 - but it looks like that PR is closed and I don't see your comment.

Side-note, should this be re-labeled as a bug instead of a question?

**Comment by XVincentX** [comment]
Re-labeled as a bug — I was trying to hack something together in 1099 but I think we should pursue another approach. I'll let you know when I have something ready to try :)

**Comment by XVincentX** [comment]
The work for this bug is tracked here: https://github.com/stoplightio/prism/pull/1101/files

This change has a good opportunity of unifying some of the things that we're doing in Prism and cleanup some stuff. It might require just a little bit more time but totally worth it.

**Comment by XVincentX** [comment]
@TylerNielsen Can you try #1101 and see if it solves your problem? I haven't finished up with that branch yet but the feature should be there.

**Comment by TylerNielsen** [comment]
@XVincentX - This is working for the original case where I noticed the issue. Thank you for the fix!
